api/try_in_order_login_provider.go

```diff
@@ -29,9 +29,12 @@ type tryInOrderLoginProviders struct {
 // Login implements the LoginProvider.Login method.
 func (p *tryInOrderLoginProviders) Login(ctx context.Context, caller base.APICaller) (*LoginResultParams, error) {
 	var lastError error
-	for _, provider := range p.providers {
+	for i, provider := range p.providers {
 		result, err := provider.Login(ctx, caller)
-		if err == nil {
+		if err != nil {
+			logger.Debugf("login error using provider %d - %s", i, err.Error())
+		} else {
+			logger.Debugf("successful login using provider %d", i)
```

Comment:
I'm not sure what `i` gives you here? 

I think a more beneficial output would be: 

```
logger.Debugf("login error using provider %s - %s", provider.Name(), err.Error())
```

Reply:
I agree, but I don't think the provider interface warrants an extra method just for this (at least not yet). At least `i` gives you some ordering info that is distinct from the order presented in the logs so you can refer to the code to see what order the providers were registered.